src/main/scala/com/campudus/tableaux/database/DatabaseConnection.scala

```diff
@@ -19,40 +19,51 @@ class DatabaseConnection(verticle: Verticle) {
 
   case class Transaction(msg: Message[JsonObject]) {
 
-    def query(query: String, values: JsonArray): Future[(Transaction, JsonObject)] = transactionHelper(Json.obj(
+    def query(query: String, values: JsonArray): Future[(Transaction, JsonArray)] = transactionHelper(Json.obj(
       "action" -> "prepared",
       "statement" -> query,
-      "values" -> values)) map { r => (Transaction(r), r.body()) }
+      "values" -> values)) flatMap { r => try Future(Transaction(r), resultChecker(r.body())) catch recover() }

```

Comment:
I think you need `Future.successful` and not `Future.apply` here. `Future.apply` will try / catch the exception itself and therefore not rollback (the recover case). You could also be able to just do `Future(Transaction(r), resultChecker(r.body()) recoverWith recover()`. I think we should rename `recover()` to something like `rollbackAndFail()` to distinguish it from the `recover` method on `Future`, too. :)